css(add): contrast-color() function page (#39847)
mdn/content: files/en-us/web/css/color_value/contrast-color/index.md

The **`contrast-color()`** functional notation takes a {{cssxref("color_value","color")}} value and returns a [guaranteed](https://www.w3.org/TR/WCAG21/#contrast-minimum) contrasting color. The function makes it easy to specify text color when background colors are created dynamically. It saves developers from maintaining background-text color pairs. It is not limited to deriving text colors; you can use it for deriving colors of anything, such as border, background, etc.

The function produces only `white` or `black` value, depending on which value produces maximum contrast with the input color. If both white and black colors produce the same contrast, then the `white` value is returned.

> [!WARNING]
> There is no guarantee that the resulting colors obtained using the `contrast-color()` function will always be accessible. For example, generally, mid-tone background colors don't result in enough contrast. It is recommended to use clearly-light or clearly-dark colors with the `contrast-color()` function.

## Syntax

```css
contrast-color(red)
contrast-color(var(--backgroundColor))
```

### Values

- `color`
  - : Any valid {{cssxref("&lt;color&gt;")}}.

> ### Examples

Reviewer: "Description" goes here, above "Examples". I'd be really interested in seeing a bit more information here about what color contrast test it uses, how it relates to WCAG, and if there are any plans to expand the functionality to generate other colors in the future, perhaps based on some kind of color wheel algorithm. To me, this seems of quite limited use considering how precise some teams' design guidelines are, but I'd be happy to hear more about the motivations for the feature design as it stands and what future plans are.

Author: The algorithm is left to browser implementations. Regarding the future, the function has undergone significant changes, so it is not wise to speculate.

Reviewer: OK

#### Contrasting text for a button

> In the following example when you change background color of the button, the webbrowser automatically applies a contrasting color to the `Submit` text.

Reviewer: ```suggestion
In the following example, the browser automatically applies a contrasting color to the submit {{htmlelement("button")}} text when you change its background color.
```

```html hidden
<label>
  Select background color of the button:
  <input type="color" id="colorPicker" value="#660066" />
</label>
<br />
<button>Submit</button>
```

```css
:root {
  --button-color: lightblue;
}

button {
  background-color: var(--button-color);

>   /* set contrasting text color automatically */

Reviewer: ```suggestion
  /* Set contrasting text color automatically */
```

  color: contrast-color(var(--button-color));
}
```

```css hidden
body {
  padding: 1rem;
}

button {
  margin: 3rem;
  padding: 1rem;
  width: 150px;
  font-size: 2rem;
  border-radius: 1rem;
}

@supports not (color: contrast-color(red)) {
  body::before {
    content: "Your browser doesn't support the contrast-color() function.";
    background-color: wheat;
    display: block;
    width: 100%;
    text-align: center;
  }

  body > * {
    display: none;
  }
}
```

```js hidden
const colorPicker = document.getElementById("colorPicker");
const root = document.documentElement;

function updateColor(color) {
  root.style.setProperty("--button-color", colorPicker.value);
}

colorPicker.addEventListener("change", updateColor);
updateColor();
```

{{EmbedLiveSample("Contrasting text for a button", "", 250)}}

#### Light and dark mode usage

> In the following example [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) [media query](/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) has been used to set background color based on operating system or browser color scheme/theme setting. The `contrast-color()` function sets text color automatically.

Reviewer: ```suggestion
In the following example, the [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) [media query](/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) has been used to set a background color based on operating system or browser color scheme settings. The `contrast-color()` function is used to automatically set the text color.
```

> In the browser setting try changing the dark mode setting to see the effect.

Reviewer: ```suggestion
Try changing the browser or OS dark mode setting to see the effect.
```

```html hidden
<pre>
    Q: How does CSS transform light into energy?
  Ans: Using <a href="/en-US/docs/Web/CSS/font-synthesis">font-synthesis</a>.
</pre>
```

```css
:root {
  --background-color: navy;
}

@media (prefers-color-scheme: light) {
  :root {
    --background-color: wheat;
  }
}

body,
a {
  background-color: var(--background-color);
  color: contrast-color(var(--background-color));
}
```

```css hidden
body {
  padding: 2rem;
  font-size: 1.2rem;
}

pre {
  margin-top: 3rem;
}

@supports not (color: contrast-color(red)) {
  body::before {
    content: "Your browser doesn't support the contrast-color() function.";
    background-color: wheat;
    display: block;
    width: 100%;
    text-align: center;
  }

  body {
    background-color: white;
  }

  body > * {
    display: none;
  }
}
```

{{EmbedLiveSample("Light and dark mode usage", "", 250)}}

> ## Specifications

Reviewer: Awesome, nice one!

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [How to have the browser pick a contrasting color in CSS](https://webkit.org/blog/16929/contrast-color/)

Reviewer: These should be rearranged a bit so that the MDN pages appear first, followed by the external pages. We usually put the ref pages first, followed by the guides, and try to get put each category in some kind of alphabetical order where possible.

Also, on external links, indicate the source and date of each one, for example:

> - [How to have the browser pick a contrasting color in CSS](https://webkit.org/blog/16929/contrast-color/) on webkit.org (2025)

---
title: contrast-color()
slug: Web/CSS/color_value/contrast-color
page-type: css-function
status:
  - experimental
browser-compat: css.types.color.contrast-color
---

{{CSSRef}}{{SeeCompatTable}}

The **`contrast-color()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Values_and_Units/CSS_Value_Functions) takes a {{cssxref("color_value","color")}} value and returns a [guaranteed](https://www.w3.org/TR/WCAG21/#contrast-minimum) contrasting color.

> `contrast-color()` makes it easy, for example, to specify a text color when background colors are created dynamically, or vice versa. It avoids the need to maintain background-text color pairs.

Reviewer: ```suggestion
`contrast-color()` makes it easy, for example, to specify a text color and automatically generate a contrasting background color, or vice versa. It avoids the need to maintain background-text color pairs.
```

## Syntax

```css
contrast-color(red)
contrast-color(var(--backgroundColor))
```

### Parameters

- `color`
  - : Any valid {{cssxref("&lt;color&gt;")}} value.

### Return value

> Either `white` or `black` color value.

Reviewer: ```suggestion
A {{cssxref("named-color")}} of either `white` or `black`.
```

## Description

> The function produces only `white` or `black` value, depending on which value produces maximum contrast with the input color. If both white and black colors produce the same contrast, then the `white` value is returned.

Reviewer: ```suggestion
The `contrast-color()` function returns a value of `white` or `black`, depending on which value has the greatest contrast with the input color. If both `white` and `black` have the same contrast with the input color, `white` is returned.
```

```html hidden
<label>
  Select background color of the button:
  <input type="color" id="colorPicker" value="#660066" />
</label>
<br />
<button>Submit</button>
```

```css
:root {
  --button-color: lightblue;
}

button {
  background-color: var(--button-color);

  /* Set contrasting text color automatically */
  color: contrast-color(var(--button-color));
}
```

```css hidden
body {
  padding: 1rem;
}

button {
  margin: 3rem;
  padding: 1rem;
  width: 150px;
  font-size: 2rem;
  border-radius: 1rem;
}

@supports not (color: contrast-color(red)) {
  body::before {
    content: "Your browser doesn't support the contrast-color() function.";
    background-color: wheat;
    display: block;
    width: 100%;
    text-align: center;
  }

  body > * {
    display: none;
  }
}
```

```js hidden
const colorPicker = document.getElementById("colorPicker");
const root = document.documentElement;

function updateColor(color) {
  root.style.setProperty("--button-color", colorPicker.value);
}

colorPicker.addEventListener("change", updateColor);
updateColor();
```

{{EmbedLiveSample("Contrasting text for a button", "", 250)}}

### Light and dark mode usage

> In the following example, the [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) [media query](/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) has been used to set a background color based on operating system or browser color scheme settings. The `contrast-color()` function is used to automatically set the text color.

Reviewer: ```suggestion
In the following example, the [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) [media query](/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) is used to set a background color based on operating system or browser color scheme settings. The `contrast-color()` function is used to set the text color automatically.
```

page-type: css-function
status:
  - experimental
browser-compat: css.types.color.contrast-color
---

{{CSSRef}}{{SeeCompatTable}}

The **`contrast-color()`** functional notation takes a {{cssxref("color_value","color")}} value and returns a [guaranteed](https://www.w3.org/TR/WCAG21/#contrast-minimum) contrasting color. The function makes it easy to specify text color when background colors are created dynamically. It saves developers from maintaining background-text color pairs. It is not limited to deriving text colors; you can use it for deriving colors of anything, such as border, background, etc.

The function produces only `white` or `black` value, depending on which value produces maximum contrast with the input color. If both white and black colors produce the same contrast, then the `white` value is returned.

> [!WARNING]
> There is no guarantee that the resulting colors obtained using the `contrast-color()` function will always be accessible. For example, generally, mid-tone background colors don't result in enough contrast. It is recommended to use clearly-light or clearly-dark colors with the `contrast-color()` function.

## Syntax

```css

> color-contrast(red)
> color-contrast(var(--backgroundColor))

Reviewer 2: ```suggestion
contrast-color(red)
contrast-color(var(--backgroundColor))
```

```

### Values

- `color`
  - : Any valid {{cssxref("&lt;color&gt;")}}.

### Examples

#### Contrasting text for a button

> ```html hidden live-sample__button_text_ex

Reviewer 2: You are not excluding any code block here, so there's no need to use `live-sample___`. Just use the section heading

<button>Submit</button>
```

```css live-sample__button_text_ex
:root {
  --button-color: lightblue;
}

button {
  background-color: var(--button-color);

  color: contrast-color(var(--button-color));
}
```

```css hidden live-sample__button_text_ex

> @supports not (color: color-contrast(red)) {
>   body::before {
>     content: "Your browser doesn't support the color-contrast() function.";

Reviewer 2: ```suggestion
@supports not (color: contrast-color(red)) {
  body::before {
    content: "Your browser doesn't support the contrast-color() function.";
```
